agent/command/git.go

```diff
@@ -94,9 +104,16 @@ type cloneOpts struct {
 	dir                    string
 	token                  string
 	recurseSubmodules      bool
+	cloneAllBranches       bool
+	fullTreeClone          bool
 	useVerbose             bool
 	usePatchMergeCommitSha bool
 	cloneDepth             int
+
+	// Since some distros or situations might cause a filtered clone of a
+	// single branch to fail, on retries we fall back to a full clone.
+	// This affects GH merge queue and PR tasks.
```

Comment:
What do you mean by this last part, "This affects GH merge queue and PR tasks"? Do other tasks not retry, or do they not use these new flags?

Reply:
I made it more clear, it affects other tasks but it has an exception for those two- which is what i meant to say

Reply:
Do we know the cases where this will definitely fail where we can just set the fallback at the top? (I'd also expected the fallback to be temporary / removed after this code is merged but do we want to keep it full time instead? In which case instead of having a command fallbackToFullClone, should we just set the new flags to true?

Reply:
cc @ZackarySantana I think this comment got missed -- do we need a fallback flag? Or could we just set the two flags to true? (i.e. should that be the same behavior as falling back to old logic?)

Reply:
The cases in which it fallbacks now are old tasks/patches and forks of private repos. With time, eventually all recent tasks will have the info and that will be solved, and I have a thread open with GitHub support and our IT about forking private repos and what is needed for it.

The only concern then with forking private repos is that it may again need to fallback when we switch from using an oauth token to a dynamic access token. 

The cases where it will fallback because it is old tasks are [here](https://github.com/evergreen-ci/evergreen/pull/7748/files#diff-1e8d12ae20a5922169a78aaae96aab11adb323b77baf8910c3cbfba80b1d0012R330) and [here](https://github.com/evergreen-ci/evergreen/pull/7748/files#diff-1e8d12ae20a5922169a78aaae96aab11adb323b77baf8910c3cbfba80b1d0012R338). For forks, there's no special logic to detect it, although it would be mostly possible since we know all approved Evergreen orgs for repos, but it doesn't cost much to just automatically fallback from it failing the first time with no permission, and ideally it will be solved in the open threads I have.

Reply:
I did some more testing and investigating, and it should also fallback for CLI patches for now. For CLI patches, this adds a small amount of time since it makes two connections to GitHub since the clone is not the final target.

Once DEVPROD-3740 is done, we can include the Git tracking org/repo + branch in the patch. Right now we include the Evergreen tracking org/repo + branch in the patch, and then apply the Git patch changes. With the Git tracking info, we can make a direct clone and only apply uncommitted changes rather than a list of commits like we do now. There's some edge cases like if they're out of sync or haven't pushed for some reason, but that should be handled in that ticket and an appropriate fallback should apply for those.

And then once we use the Git tracking details, we should remove the fallback for CLI patches. I'll make a follow up ticket that depends on DEVPROD-3740 and add the fallback for CLI patches.

Reply:
I made: [DEVPROD-6795](https://jira.mongodb.org/browse/DEVPROD-6795)

Reply:
[This](https://spruce-staging.corp.mongodb.com/task/evg_ubuntu2204_test_db_patch_930ce9bc83f06a4ff555645dfcae46813463ea8a_6629104cfb03b80007cdf51a_24_04_24_13_59_52/logs?execution=2&logtype=all&sortBy=STATUS&sortDir=ASC) is a task showing the fallback working (in the parsely logs, you can see it doesn't use the new flags. If you look at previous executions, before the fallback, it does use the flags- and takes a little longer because it has to do a 2nd lookup/fetch)